SequiturTool.py

```diff
@@ -54,17 +54,32 @@ def __call__(self, log, context, model):
 def transposeSample(sample):
     return [ (right, left) for left, right in sample ]
 
-def partitionSample(sample, portion = 0.1):
-    trainSample = []
-    develSample = []
+def partition_sample(sample, portion = 0.1):
+    train_sample = []
+    devel_sample = []
     j = 0
-    for i, s in enumerate(sample):
-        if j / (i+1) < portion:
-            develSample.append(s)
+    for i, s in enumerate(group_by_orth(sample)):
+        if j / (i + 1) < portion:
+            for value in s[1]:
+                devel_sample.append((s[0], value))
             j += 1
         else:
-            trainSample.append(s)
-    return trainSample, develSample
+            for value in s[1]:
+                train_sample.append((s[0], value))
+    return train_sample, devel_sample
+
+
+def group_by_orth(sample):
+    import random
+
+    source_values = []
+    mapping = defaultdict(list)
+    for s in sample:
+        source_values.append(s[0])
```

Comment:
this is kinda weird -- your list will continue duplicate entries so the new selection probability wont be 1/no_words but 1/no_lines (which is the same as the old probability, but in this setting it will make some words more probable to be selected, which I think we don't want)

Reply:
ok, I think I wasn;t mistaken -- I think you want to write something like (in diff form)
```
 def group_by_orth(sample):
     import random

-    source_values = []
+    source_values = set()
     mapping = defaultdict(list)
     for s in sample:
-        source_values.append(s[0])
+        source_values.add(s[0])
         mapping[s[0]].append(s[1])
+    source_values = list(source_values)
     random.shuffle(source_values)
     for s in source_values:
         yield s, mapping[s]
```
The reason is that doing it your way, you _will_ draw the same graphemic form from source_values list (because it's a list) several times. So you will end up with duplicated additions into train_sample or devel_sample and you might even end up with the word and all it's prons in both sets